src/core/magnetostatics/dds.cpp

```diff
@@ -30,110 +30,376 @@
+ * 5. Calculate the interaction with the rest of the particles. Here
+ *    every pair is visited twice (not necessarily on the same rank)
+ *    so that no reduction of the forces is needed.
+ *
+ * Logically this is equivalent to the potential calculation
+ * in @ref DipolarDirectSum::long_range_energy, which calculates
+ * a naive N-square sum, but has better performance and scaling.
+ */
+void DipolarDirectSum::add_long_range_forces(
+    ParticleRange const &particles) const {
+  auto const &box_l = ::box_geo.length();
+  /* collect particle data */
+  std::vector<Particle *> local_particles;
+  std::vector<PosMom> local_posmom;
+  std::vector<PosMom> all_posmom;
+  std::vector<boost::mpi::request> reqs;
+  int offset{0};
+  std::tie(local_particles, local_posmom, all_posmom, reqs, offset) =
+      gather_particle_data(particles, n_replicas);
+
+  /* Number of image boxes considered */
+  auto const ncut = get_n_cut(n_replicas);
+  auto const with_replicas = (ncut.norm2() > 0);
+
+  /* Range of particles we calculate the ia for on this node */
+  auto const local_posmom_begin = all_posmom.begin() + offset;
+  auto const local_posmom_end = local_posmom_begin + local_particles.size();
+
+  /* Output iterator for the force */
+  auto p = local_particles.begin();
+
+  /* IA with local particles */
+  for (auto pi = local_posmom_begin; pi != local_posmom_end; ++pi, ++p) {
+    /* IA with own images */
+    auto fi = image_sum(
+        pi, std::next(pi), pi, with_replicas, ncut, box_l, ParticleForce{},
+        [pi](Utils::Vector3d const &rn, Utils::Vector3d const &mj) {
+          return pair_force(rn, pi->m, mj);
+        });
+
+    /* IA with other local particles */
+    auto q = std::next(p);
+    for (auto pj = std::next(pi); pj != local_posmom_end; ++pj, ++q) {
+      auto const d = (with_replicas)
+                         ? (pi->pos - pj->pos)
+                         : ::box_geo.get_mi_vector(pi->pos, pj->pos);
+
+      ParticleForce fij{};
+      ParticleForce fji{};
+      for_each_image(ncut, [&](int nx, int ny, int nz) {
+        auto const rn =
+            d + Utils::Vector3d{nx * box_l[0], ny * box_l[1], nz * box_l[2]};
+        auto const pf = pair_force(rn, pi->m, pj->m);
+        fij += pf;
+        fji.f -= pf.f;
+        /* Conservation of angular momentum mandates that
+         * 0 = t_i + r_ij x F_ij + t_j */
+        fji.torque += vector_product(pf.f, rn) - pf.torque;
+      });
+
+      fi += fij;
+      (*q)->f.f += prefactor * fji.f;
+      (*q)->f.torque += prefactor * fji.torque;
     }
+
+    (*p)->force() += prefactor * fi.f;
+    (*p)->torque() += prefactor * fi.torque;
   }
 
-  return energy;
+  /* Wait for the rest of the data to arrive */
+  boost::mpi::wait_all(reqs.begin(), reqs.end());
+
+  /* Output iterator for the force */
+  p = local_particles.begin();
+
+  /* Interaction with all the other particles */
+  for (auto pi = local_posmom_begin; pi != local_posmom_end; ++pi, ++p) {
+    // red particles
```

Comment:
For interactions among local particles, each pair is only looked at once and symmetry is used to calculate the oposite force/torque. For interactions between particles at different nodes, this seems not to be the case, so f_ij and f_ji are calculated separately. the result is correct. However, in the future, this could/should be expanded to use the same symmetry and have a final reduciton of the forces. Due to the n^2 scaling, this is probably worth it. Not now, though.